doc/syslog/python-logger-enhancement.md

```diff
@@ -0,0 +1,217 @@
+
+- A class level logger registry shall be added to `SysLogger`class
+- Each logger instance shall register itself to logger register if enables runtime configuration
+- Logger configuration shall be refreshed by CLI which send a SIGHUP signal to the daemon
+
+![python-logger-enhancement](/doc/syslog/images/python_logger_enhancement.svg)
+
+### High-Level Design
+
+#### SysLogger class change
+
+- A new class level variable `log_registry` shall be added to `SysLogger`class. The type is `dict`.
+- A new argument `enable_runtime_config` shall be added to `SysLogger.__init__`. Default to False.
+- SysLogger instance shall register itself to `log_registry` if `enable_runtime_config` is True.
+- SysLogger instance shall read log level configuration from CONFIG DB if `enable_runtime_config` is True and DB entry is available.
+- SysLogger instance shall update its log level configuration to CONFIG DB if `enable_runtime_config` is True and DB entry is not available.
+- A new class level method `refresh_config` shall be added to `SysLogger` class. The method shall be called while receiving SIGHUP to update log configuration for each logger instance registered to `log_registry`.
+
+SysLogger create flow:
+
+![python-logger-create-flow](/doc/syslog/images/python_logger_create_flow.svg)
+
+Refresh configuration flow:
+
+![python-logger-refresh-config-flow](/doc/syslog/images/python_logger_refresh_config_flow.svg)
+
+#### Multi threading support
+
+- `SysLogger` shall use `threading.Lock` to protect `log_registry`
+- Multi-threading support is transparent to daemons who uses `SysLogger`
+
+#### Multi processing support
+
+For a multi-processing daemon, it shall do follow things to support this feature:
+
+1. Child process shall create logger instance in its own context
+
+Wrong way:
+```python
+log = syslogger.SysLogger('daemon_name', enable_runtime_config=True)
+
+def run():
+    while True:
+        # log instance is cloned from main process, it is not registered to the global
+        # log registry of child process
+        log.log_notice('log something')
+
+p = multiprocessing.Process(target=run)
+p.join()
+```
+
+Right way:
+```python
+log = syslogger.SysLogger('daemon_name', enable_runtime_config=True)
+
+def child():
+    child_log = syslogger.SysLogger('daemon_name', enable_runtime_config=True)
+    while True:
+        # child_log is created in child process, so it will register itself to
+        # the global log registry of child process
+        child_log.log_notice('log something')
+
+p = multiprocessing.Process(target=child)
+p.join()
+```
+
+2. Main process shall forward SIGHUP to child processes. There are many options to achieve this, here is an example:
```

Comment:
@Junchao-Mellanox Why we need per process handling? The signal handling is per process so child process is independ of parent process and can have its own signal handling. Probably what we need is to support single process with multiple threads.

Reply:
Hi @prgeor , this is just an example in case user uses it in multi-processing case. Your suggestion is also a good option and probably a more general option for multi processing case. Basically, this feature supports single process with multiple threads well.

Reply:
fixed